### DISCUSSION TITLE
How to create migrations with specific pgvector-enabled docker image?

### DISCUSSION THREAD

**Discussion by leomorpho** [ariga/atlas#3013] [Q&A] [answered]
I am puzzled as to how to solve my issue. I am currently creating my migrations with the following command:

```
atlas migrate diff "$(name)" \
	  --dir "file://ent/migrate/migrations" \
	  --to "ent://ent/schema" \
	  --dev-url "docker://postgres/15/test?search_path=public"
```

However I recently added pgvector, and tried changing the temporary dev container to use pgvector with no success.
I tried doing:
```
atlas migrate diff "$(name)" \
	  --dir "file://ent/migrate/migrations" \
	  --to "ent://ent/schema" \
	  --dev-url "docker+postgres://ghcr.io/ankane/pgvector:v0.5.1/dev?search_path=public"
```
which fails with
```
Error: sql/sqlclient: unknown driver "docker+postgres". See: https://atlasgo.io/url
```

And I tried with a locally-running DB, but attempting to create the migrations always yields `Error: sql/migrate: connected database is not clean`, which I kind of expected.

I tried many other things from https://atlasgo.io/concepts/dev-database#diffing to no avail.
What am I missing, and how can I create migrations using a pgvector-enabled dev database?

Thank you kindly for any assistance provided.

**Comment by rotemtam** [comment]
What version of Atlas are you using ?  (Run "atlas version")

Also, you may find [this guide](https://entgo.io/docs/migration/extensions) from the Ent website helpful

**Reply by leomorpho** [reply]
I am using
```sh
❯ atlas version
atlas version v0.15.1-27ac69a-canary
https://github.com/ariga/atlas/releases/latest
```
Thanks for the guide link, I will go through that.

**Reply by a8m** [reply]
Let us know if you still have troubles with it, and we’ll try to reproduce this.

**Reply by rotemtam** [reply]
> I am using
> 
> ```sh
> 
> ❯ atlas version
> 
> atlas version v0.15.1-27ac69a-canary
> 
> https://github.com/ariga/atlas/releases/latest
> 
> ```
> 
> Thanks for the guide link, I will go through that.

An ancient version! Please upgrade and try again:

curl -sSf https://atlasgo.sh | sh

**Reply by leomorpho** [reply, marked as answer]
Thanks, yes, it was an ancient version and I did not realize...my bad.

Sharing a solution I used to create migrations against a pgvector-enabled image.

In my makefile:
```Make
# Define variables
PGVECTOR_IMAGE_NAME = custom-pgvector-for-atlas
PGVECTOR_IMAGE_TAG = latest
PGVECTOR_IMAGE_DIR = pgvector-image

.PHONY: migrate_diff
makemigrations: build-image ## Create a migration
	@echo "Running Atlas migrate diff..."
	atlas migrate diff "$(name)" \
	  --dir "file://ent/migrate/migrations" \
	  --to "ent://ent/schema" \
	  --dev-url "docker+postgres://_/custom-pgvector:latest/dev?search_path=public&sslmode=disable"
```

And then in `pgvector-image/Dockerfile`:
```Dockerfile
# Use the pgvector base image
FROM ankane/pgvector:v0.5.1

# Set environment variables for PostgreSQL
ENV POSTGRES_USER=postgres
ENV POSTGRES_DB=postgres
ENV POSTGRES_HOST_AUTH_METHOD=trust

# Add initialization script
COPY init_pgvector.sh /docker-entrypoint-initdb.d/

# Ensure the script has the correct permissions
RUN chmod +x /docker-entrypoint-initdb.d/init_pgvector.sh
```
and in `pgvector-image/init_pgvector.sh`:
```sh
#!/bin/bash
set -e

# Function to check if PostgreSQL is ready
check_postgres() {
  pg_isready -U "$POSTGRES_USER" -d "$POSTGRES_DB"
}

# Wait for PostgreSQL to be ready
echo "Waiting for PostgreSQL to start..."
until check_postgres; do
  echo "PostgreSQL is unavailable - sleeping"
  sleep 2
done

# Create the extension
echo "PostgreSQL is up - creating extension"
psql -U "$POSTGRES_USER" -d "$POSTGRES_DB" -c "CREATE EXTENSION IF NOT EXISTS vector;"
```

That ended up being a little more complicated than expected to get it to work, but now working just fine with pgvector ❤️
